### PR TITLE
Remaining tx ridigity issues

### PR DESCRIPTION
**Pull request by jeffro256**
Use free functions to do serialization in-place. Simplifies code, reduces binary size, and reduces copying.

### CODE REVIEW

**Comment by UkoeHB** [comment]
This PR is a regression without compact serialization. The existing implementation is specifically geared toward the special needs of compactly serializing and deserializing a transaction.

I cannot merge this until/unless you show how to do it without regressing.

**Comment by jeffro256** [comment]
Yeah I can definitely make it more compact, but I'll wait to see if we like #31 before refactoring this PR to make it more compact.

**Comment by jeffro256** [comment]
Ok I completely shifted the focus of this PR: it also fixes the remaining tx class format issues as I see them. Those being: the extra key image in `rct::clsag` and implied bin config and seed generators in the membership proofs. It depends on the bulletproof  change PR and the free function serialization PR.

**Comment by jeffro256** [comment]
I also have a plan to make the serialization much more compact by removing almost all length serializations, but that can come in a later PR as this one is already getting big.

File: src/serialization/serialization.h
Comment (UkoeHB) [inline src/serialization/serialization.h]: Move these serialization additions into a new PR. This also needs a direct Monero repo PR (everything in this repo ultimately has to be merged to Monero).

File: tests/performance_tests/rct_scalarmult8.h
Comment (UkoeHB) [inline tests/performance_tests/rct_scalarmult8.h]: Move this perf test to a new PR.

File: src/seraphis_crypto/sp_legacy_proof_helpers.cpp
Comment (UkoeHB) [inline src/seraphis_crypto/sp_legacy_proof_helpers.cpp:82]: Why is this being removed?

**Review by UkoeHB — CHANGES_REQUESTED** [review]
Brief comments.

File: src/serialization/serialization.h
Comment (jeffro256) [reply src/serialization/serialization.h]: These changes are already PR'ed to Monero core: https://github.com/monero-project/monero/pull/9194

File: tests/performance_tests/rct_scalarmult8.h
Comment (jeffro256) [reply tests/performance_tests/rct_scalarmult8.h]: They are in the BP PR, but I can move to a new, separate PR

File: src/seraphis_crypto/sp_legacy_proof_helpers.cpp
Comment (jeffro256) [reply src/seraphis_crypto/sp_legacy_proof_helpers.cpp:82]: `LegacyClsagProof` replaces `rct::clsag` as a transaction component. Only difference is no `I` (key image) field since already stored in legacy input image.

**Comment by jeffro256** [comment]
Rebased. Will split into multiple PRs here soon

**Comment by jeffro256** [comment]
Okay I pulled out everything from this PR except the stuff that is unique to it: AKA the new `LegacyClsagProof`, refactorings of `SpMembershipProofV1`, implying seraphis ref set config based on tx semantic rules version, as well as the validation context changes that go along with that.

File: src/seraphis_impl/serialization_demo_utils.cpp
Comment (UkoeHB) [inline src/seraphis_impl/serialization_demo_utils.cpp]:
```suggestion
void make_serializable_clsag(const LegacyClsagProof &clsag, ser_LegacyClsagProof &serializable_clsag_out)
```

File: src/seraphis_impl/serialization_demo_utils.cpp
Comment (UkoeHB) [inline src/seraphis_impl/serialization_demo_utils.cpp:295]: This is a change because previously it was serializing as offsets with `indices_to_offsets` and `indices_from_offsets`. I'm not actually sure if the old version or the plain 2-byte version is cheaper. It depends how wide the bins are.

File: src/seraphis_main/tx_builders_inputs.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_builders_inputs.cpp]:
```suggestion
    membership_proof_out.bin_loci            = std::move(binned_reference_set.bin_loci);
    membership_proof_out.bin_rotation_factor = binned_reference_set.bin_rotation_factor;
```

File: src/seraphis_main/tx_builders_legacy_inputs.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_builders_legacy_inputs.cpp:326]: The disadvantage is you have to allocate to convert it.

File: src/seraphis_main/tx_builders_mixed.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_builders_mixed.cpp]:
```suggestion
// - stores given seraphis reference set config
```

File: src/seraphis_main/tx_builders_mixed.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_builders_mixed.cpp]:
```suggestion
    virtual SemanticConfigSpRefSetV1& get_sp_ref_set_config(const tx_version_t) const override
```

File: src/seraphis_main/tx_validation_context.h
Comment (UkoeHB) [inline src/seraphis_main/tx_validation_context.h]:
```suggestion
    virtual const SemanticConfigSpRefSetV1& get_sp_ref_set_config(const tx_version_t txver) const = 0;
```
Use a ref for > 16 bytes (2 pointers).

File: src/seraphis_main/tx_builders_mixed.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_builders_mixed.cpp]:
```suggestion
    return SemanticConfigSpRefSetV1{.decomp_n = 2, .decomp_m = 2, .bin_radius = 0, .num_bin_members = 1};
```

File: src/seraphis_main/tx_builders_mixed.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_builders_mixed.cpp]:
```suggestion
    const TxValidationContextSimple tx_validation_context{
        sp_ref_set_config,
```

File: src/seraphis_main/tx_component_types_legacy.h
Comment (UkoeHB) [inline src/seraphis_main/tx_component_types_legacy.h]:
```suggestion
    /// scalars
    rct::keyV s;
    rct::key c1;
```

File: src/seraphis_main/tx_validators.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_validators.cpp]:
```suggestion
    if ((ref_set_size/num_bin_loci)*num_bin_loci != ref_set_size)
        return false;
```
Isn't modulo kind of expensive?

File: src/seraphis_main/tx_validators.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_validators.cpp]:
```suggestion
        if (sp_proof.bin_rotation_factor >= bin_width)
            return false;
```

File: src/seraphis_main/tx_validators.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_validators.cpp]:
```suggestion
        if (!std::is_sorted(sp_proof.bin_loci.begin(), sp_proof.bin_loci.end()))
```

File: src/seraphis_main/tx_validators.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_validators.cpp]:
```suggestion
        if(!try_get_reference_indices_from_binned_reference_set_v1(binned_reference_set, reference_indices))
```

File: src/seraphis_main/txtype_squashed_v1.cpp
Comment (UkoeHB) [inline src/seraphis_main/txtype_squashed_v1.cpp]:
```suggestion
    // subtract balance proof size and add its weight
    weight -= sp_balance_proof_v1_size_bytes(num_sp_inputs + num_outputs);
    weight += sp_balance_proof_v1_weight(num_sp_inputs + num_outputs);

```

File: src/seraphis_main/txtype_squashed_v1.cpp
Comment (UkoeHB) [inline src/seraphis_main/txtype_squashed_v1.cpp]:
```suggestion
    const SemanticConfigSpRefSetV1 sp_ref_set_config{tx_validation_context.get_sp_ref_set_config(
        tx_version_from(tx_semantic_rules_version)
    )};
```

File: src/seraphis_mocks/mock_send_receive.cpp
Comment (UkoeHB) [inline src/seraphis_mocks/mock_send_receive.cpp]:
```suggestion
        SemanticConfigSpRefSetV1{
            .decomp_n = ref_set_decomp_n,
            .decomp_m = ref_set_decomp_m,
            .bin_radius = bin_config.bin_radius,
            .num_bin_members = bin_config.num_bin_members
        }
```

File: src/seraphis_mocks/mock_send_receive.cpp
Comment (UkoeHB) [inline src/seraphis_mocks/mock_send_receive.cpp]:
```suggestion
        SemanticConfigSpRefSetV1{
            .decomp_n = ref_set_decomp_n,
            .decomp_m = ref_set_decomp_m,
            .bin_radius = bin_config.bin_radius,
            .num_bin_members = bin_config.num_bin_members
        }
```

File: src/seraphis_mocks/mock_send_receive.cpp
Comment (UkoeHB) [inline src/seraphis_mocks/mock_send_receive.cpp]:
```suggestion
        SemanticConfigSpRefSetV1{
            .decomp_n = ref_set_decomp_n,
            .decomp_m = ref_set_decomp_m,
            .bin_radius = bin_config.bin_radius,
            .num_bin_members = bin_config.num_bin_members
        }
```

**Review by UkoeHB — CHANGES_REQUESTED** [review]
Minor comments

File: src/seraphis_impl/serialization_demo_utils.cpp
Comment (jeffro256) [reply src/seraphis_impl/serialization_demo_utils.cpp:295]: Ngl, I didn't spend much time fussing over the details of serialization since I replaced the code anyways in #39. IMO that's the PR where we should focus on compactness of the format

File: src/seraphis_main/tx_validators.cpp
Comment (jeffro256) [reply src/seraphis_main/tx_validators.cpp]: On x86 systems, all `DIV` instructions do modulo at the same time. Division is (almost always) the expensive part of modulo anyways

Comment (jeffro256) [reply src/seraphis_main/tx_validators.cpp]: For something as basic as the `%` operator, the compiler will probably optimize it in the best way

Comment (UkoeHB) [reply src/seraphis_main/tx_validators.cpp]: Ok sounds good

**Comment by jeffro256** [comment]
Rebased x2

File: src/seraphis_main/tx_validators.cpp
Comment (UkoeHB) [inline src/seraphis_main/tx_validators.cpp]:
```suggestion
    if ((num_bin_loci == 0) || (ref_set_size % num_bin_loci != 0))
```
Potential division by zero here.

**Review by UkoeHB — CHANGES_REQUESTED** [review]

**Pull request merged**
